### ISSUE TITLE
Port existing `async` package to new `Task` type

### ISSUE THREAD

**Issue by pchiusano** [functional-streams-for-scala/fs2#441]
Even though I think the existing `async` package will be less needed with the new design, I still think it's important to have ports of the existing types, especially the `Queue` and `Signal` types which are used a lot.

@pchlupacek I think you are the best person for this given how much of the `async` code you've written so would you be willing to take this on? It might be really simple, or there might be difficulties because the new `Task` doesn't support `runAsyncInterruptibly` (not sure if that is being used anywhere). But note that you can use [`cancellableGet` from a `Task.Ref[A]`][cget], if that helps.

[cget]: https://github.com/scalaz/scalaz-stream/blob/topic/redesign/src/main/scala/fs2/util/Task.scala#L307

I would work on a branch off `topic/redesign`, and place this in `fs2.unstable.async`.

**Comment by pchlupacek** [comment]
@pchiusano agreed.

**Comment by pchlupacek** [comment]
@pchiusano I think with `Task.Ref` in place we may not need that interruption altogether. I think we basically used it only in Wye and nJoin, not needed in these primitives. If I would need something more, I'll just add it to Task.Ref // Async.

**Comment by pchlupacek** [comment]
@pchiusano I have WIP here : https://github.com/scalaz/scalaz-stream/blob/topic/redesign-signal/src/main/scala/fs2/async/mutable/Signal.scala 

Before I will go further deep in implementation, I would like to clarify following: 

* With the new Async type, I think it is possible to not specialise to `Task`. The only think we need is to add to `Ref` is the following 
   - compareAndSet - this I think will be necessary to avoid race conditions anyhow
   - getAndSet 
   - register  - `def register(cb: A => Unit):F[A]`  // invoked when the next `A` is set

I think this is beneficial, hence `Signal` and `Ref` are very in common and I don't think so we need just for a very subtle differences two different implementation. 
 
For the API, I am thinking of removing close, fail and etc. I think these are not really necessary. When someone really needs the close/fail behaviour, easiest is to make `Signal[Throwable \/ A]` instead of just `[A]` if you really care about termination reason or just `Option[A]` if you don't.  I think to leave fail/close in signal, I will need to know how to terminate and correctly propagate all the exceptions and throwables through.  

Also I would like instead of Long for identities use in AsyncTask instance the `new TaskIdentity` and then do referential equality comparison instead the `Longs` passed along with `Get` 

LMK any thoughts ideas, then I will finish this implementation

**Comment by pchiusano** [comment]
+1 on removing close, fail, etc
Fine with using reference equality for Get

The other stuff I'm not sure on

compareAndSet and getAndSet seem ok, though maybe they can be defined in
terms of something simpler

Maybe modify : (A => F[A]) : F[A]

until the returned Task completes the ref just queues up other modifies

Can be used to implement the other operations

I don't like register, callbacks should not be part of the Api, only the
implementation. I think it should just be next : Ref[A]=>F[A] which behaves
like get but waits until the next modify

Will respond Monday or Tuesday when I have more time, sorry

**Comment by pchlupacek** [comment]
@pchiusano thanks 

Modify seems ok, will try to implement them this way. 

Will see if Ref[A] => F[A] will work. I think there is chance it will.

I'll try to move on a bit tomorrow...

**Comment by pchlupacek** [comment]
@pchiusano I did some experiments and so far this seems to be I will need to add to an api : 
```scala

 /**
   * Obtains the value of `Ref` if `A` satisfies `f` (Right) on read.
   * Note that if `Ref` is not yet set, it awaits until it is set for the first time.
   * Then, `f` is then consulted on any next `Set` until it is satisfied (Right).
   * In that case, the produced `A` is used to modify the `Ref` and is returned together with
   * original value passed to `f`
   *
   * If the `f` is not satisfied, it may signal on Left if other getAndSet's shall be consulted (true)
   * or if the evaluation of all other getAndSets shall terminate early (false).
   *
   * Note that any `getAndSet` received are applied to incoming `A` in same order as they have been
   * received by `Ref`
   */
  def getAndSet[A](r:Ref[A],f:A => Either[Boolean,A]): F[(A,A)]



  /**
   * like `getAndSet`, but returns an `F[Option[(A,A)]]` that can be used to cancel the subscription.
   * Note that, if resulting `F` returns None, then this guarantees that `f` was not yet evaluated.
   * Otherwise, if `F` returns Some(a,a) then `f` was evaluated and original value with result is returned.
   */
  def cancellableGetAndSet[A](r: Ref[A], f:A => Either[Boolean,A]): F[(F[(A,A)], F[Option[(A,A)]])]


```

I am quite confident that these are all what we need for `signal` and likely for the `queue` implementation as well. Unfortunately implementation of Ref[Task] with these is not that straightforward as before, as you may see in the comments there are couple of invariants we need to care in `Ref` implementation. 

What are your thoughts? Any ideas before I implement it?

**Comment by pchiusano** [comment]
Okay, I thought about this some. Here is what I am thinking, you might not like this, @pchlupacek. :)

* I'd like to just do a very conservative port of the existing API, keeping the same public API including the specialization to the new `Task` (at least for the external, publicly facing API). My reasoning is that it will be much easier and less rushed to experiment with and hash out these design changes after the 0.9 release, especially since the `async` package will be in the `unstable` repo and will have its own release cycles then. It feels like we're rushing it to and I'd rather not force a major API change on people at the same time as the other changes to the core anyway.
* For a future release, even though I like the idea of generalizing `Signal` and `Queue` to work with stuff other than `Task`, I will always want to keep the `Async` trait as is, and introduce another trait for any extra capabilities needed by the `async` package (I can't think of a better name than `AsyncExtended`). Reason is that the existing `Async` is sufficient to implement the core library, which is useful on its own, so the less that's in it, the easier it is for people to bring their own `Task`-like types.

@pchlupacek lmk if you are okay with this, or if you have objections.

**Comment by pchlupacek** [comment]
@pchiusano ok. I still would like to implement queue and signal with minimal api introductions and as such will perhaps still try to use `AsyncExt` or similar to do that task. When I will be done we'll see how that went. 

The only think that worries me is that `unstable` flag on Queue and Signal. I hardly believe that you can write any sensible program w/o them, and `unstable` tag is not what I would like to based my code on. I think these two primitives are most tested and used concurrent primitives in current code and even with introduction of new `task` combinator, I am still confident you will need theses in any program. I think marking them `unstable` is just not right and will confuse users. 

If we really planning on tossing them out and replacing them with something different, given the amount of changes fs2._ will require to the user's code anyhow, I would rather focus on new improved api than reintroduce something `just` for compatibility that would be questionable anyhow (termination causes, resource allocation rules, name it all).

**Comment by pchiusano** [comment]
@pchlupacek I hear you and I think we should give this more thought. Some comments inline below. I would say proceed with implementing queue and signal, using `AsyncExt` if you like. Once you have it working, we can decide later what public API to expose.

> The only think that worries me is that unstable flag on Queue and Signal.

I consider them "unstable" in the sense that they are subject to API changes, like what you have already been considering. Not in the sense of being potentially buggy - I think they are solid in that department. :) But maybe the name 'unstable' will be confusing. Maybe `develop` or `dev` instead.

> even with introduction of new task combinator, I am still confident you will need theses in any program.

I guess I'm not as convinced of this. The new API can express a lot. You only need `Signal` and `Queue` if you need mutable state in your network, they are no longer needed just to get concurrency. Because of this, I'd like to de-emphasize their use - if you need concurrency, use the regular API. If you need to invert control and bind to an imperative API, use the `task` module (maybe it would be a good idea to generalize this to any `F` using `AsyncExt`).

> If we really planning on tossing them out and replacing them with something different, given the amount of changes fs2._ will require to the user's code anyhow, I would rather focus on new improved api...

The thing is, even though there are a lot of changes in fs2, unless you are writing your own transducers, which many people aren't, there's a large subset of code which is source-compatible. Changing `Queue` and `Signal` breaks that. But I hear what you are saying.

**Comment by pchlupacek** [comment]
@pchiusano just give you an idea in our 100k line scalaz-stream project 
- .pipe  - 300 times
- .wye - 200 times
- Queue - 2500+ times
- Signal - 707 times
- Channel - 427 times
- Sink - 664 times

**Comment by pchlupacek** [comment]
what I wanted to say, I don't want to modify this twice, but only once :-)

**Comment by pchiusano** [comment]
@pchlupacek those numbers are insane. :) What about `Topic`?

I think there is a bit of tension here - people who are using the library heavily probably will have an easier upgrade if we do everything in one go. People who are using the library more casually and aren't writing their own transducers might have a pretty painless 0.9 upgrade if the existing async interface isn't changed as much.

I'm not sure what the best thing to do is, but I'd like to get this task unblocked. @pchlupacek I think you should implement the various `async` types as you see fit. So feel free to generalize beyond `Task` (if you do, introduce a new interface that extends `Async` rather than modifying `Async` directly), and feel free to try removing the `close` and `fail` methods (in favor of just using `Either` or whatever as the element of the `Queue`/`Signal`). You can put the definitions in the `fs2.async` package for now.

Even if the implementation is more general, it will be easy later to supply monomorphic versions if we want and/or shuffle around the package structure.

**Comment by pchlupacek** [comment]
@pchiusano got it. and topic in fact we don't use that much. We have build more complex distributed logic around it, so we have now topic that persist to database and perhaps in code we have 2-3 places where we work with it. I think we can perhaps safely remove writertopic and replace it with just topic, but need more thought on that.

**Comment by pchlupacek** [comment]
@pchiusano I don't think that this implementation of `set` is correct. I think Set task must complete AFTER it was set, otherwise there won't be evidence that set succeeded, at that can cause various coordination issues. 
Can you please comment on that? Maybe I am missing something?
https://github.com/scalaz/scalaz-stream/blob/topic/redesign/src/main/scala/fs2/util/Task.scala#L300

**Comment by pchiusano** [comment]
@pchlupacek the implementation of `set` is supposed to be asynchronous / nonblocking, I tried to sort of document that [here](https://github.com/scalaz/scalaz-stream/blob/topic/redesign/src/main/scala/fs2/Async.scala#L26) but it could probably be clearer :) 

If you want a blocking `set`, where when the returned `Task` completes the value has been updated you can define that by doing a `set` immediately followed by a `get`:

```Scala
def blockingSet[A](r: Ref[A])(t: F[A]): F[Unit] = 
  bind(set(ref)(t))(_ => map(get(ref))(_ => ()))
```

I'm okay with adding this to the `Async` trait if you want, in case instances have a more efficient implementation.

**Comment by pchlupacek** [comment]
@pchiusano thanks for suggestion. Not sure what do you mean by `set being non-blocking`, when it can be blocked? You just aiming to save this single callback ? I don't believe the `blockingSet` is correct name, as from your implementation suggestion set is non-blocking, it is fully asynchronous and only guarantees that it has been evaluated when task complete.

**Comment by pchlupacek** [comment]
@pchiusano since signal and queue is now in, There are my todo list for this, can you perhaps comment?
- introduce closeable queue and signal (defined in terms of signal and queue)
- introduce topic (I am not sure on writer topic yet, what do you think ? )
- introduce closeable topic 
- anything missing?

**Comment by pchiusano** [comment]
I'd like to punt on implementing topic (and closeable topic or writer topic) for 0.9. These need more design work and we already have a ton of stuff to do. Other than that I think that list looks good.

**Comment by mpilquist** [comment]
OK, I created new issues for the remaining pieces. Closing this now.
